### ISSUE TITLE
MethodError: Cannot `convert` an object of type GDFile

### ISSUE THREAD

**Issue by yakir12** [oxinabox/PyDrive.jl#1]
I'm getting the following errors now:
```julia
WARNING: Could not resolve filename due to
WARNING: Could not resolve filename due to
WARNING: Could not resolve filename due to
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: Could not resolve filename due to
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: Could not resolve filename due to
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
WARNING: Could not resolve filename due to
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
ERROR: WARNING: falling back to using final part of remotepath
LoadError: PyError (ccall(@pysym(:PyObject_Call), PyPtr, (PyPtr, PyPtr, PyPtr), o, arg, C_NULL)) <class 'pydrive.files.FileNotDownloadableError'>
FileNotDownloadableError('No downloadLink/exportLinks for mimetype found in metadata',)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 210, in GetContentFile
    self.FetchContent(mimetype, remove_bom)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 43, in _decorated
    return decoratee(self, *args, **kwargs)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 265, in FetchContent
    'No downloadLink/exportLinks for mimetype found in metadata')

Stacktrace:
 [1] #573 at ./asyncmap.jl:178 [inlined]
 [2] foreach(::Base.##573#575, ::Array{Any,1}) at ./abstractarray.jl:1733
 [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:178
 [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Int64, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:154
 [5] #async_usemap#558(::Int64, ::Void, ::Function, ::DataDeps.##20#21{#drive_download,String}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:103
 [6] (::Base.#kw##async_usemap)(::Array{Any,1}, ::Base.#async_usemap, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./<missing>:0
 [7] #download#19(::Array{GDFile,1}, ::Void, ::Bool, ::Function, ::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:48
 [8] handle_missing(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:6
 [9] _resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:83
 [10] resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:31
 [11] resolve(::String, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:54
 [12] resolve(::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:73
 [13] include_from_node1(::String) at ./loading.jl:576
 [14] include(::String) at ./sysimg.jl:14
while loading /home/yakir/pydrive.jl, in expression starting on line 13
```
It might have to do with [DataDeps.jl/pull/54](https://github.com/oxinabox/DataDeps.jl/pull/54) and your explanation there:
> https://discourse.julialang.org/t/datadeps-and-google-drive/12271/11?u=oxinabox
> Though that code needs to be updated to to return the filename.
> Which is actually a trivial change since the filename is exposed as metadata(gdfile, "title")

So what should we change exactly?

**Comment by oxinabox** [comment]
54 has been merged but not yet tagged.
So unless you've checked out DataDeps master,
It should be not changed.

I will look into this

**Comment by yakir12** [comment]
It might be something I'm doing wrong. hmmm...

**Comment by yakir12** [comment]
Hmm, I don't think it's me, because it is getting the correct file names from the desired folder:
```julia
INFO: This program has requested access to the data dependency GoogleDriveDemo.
INFO: which is not currently installed. It can be installed automatically, and you will not see this message again.
INFO: 
Demonstration of google drive

INFO: Do you want to download the dataset from GDFile[Google Drive file: Budget, Google Drive file: Bordsplacering, Google Drive file: Inköp, Google Drive file: Inköpslista:, Google Drive file: Ansvarsfördelning, Google Drive file: Lyrics Sånghäfte, Google Drive file: Skylt2.pdf, Google Drive file: Skylt.pdf, Google Drive file: Guest List, Google Drive file: Dekoration, Google Drive file: Ansvarsfördelning mat, Google Drive file: Ceremoni, Google Drive file: Adresslista, Google Drive file: Guest List with colors, Google Drive file: Invitation, Google Drive file: Copy of berg-bengtsson, Google Drive file: kommer ni eller inte?!, Google Drive file: IMG_7020.psd, Google Drive file: IMG_7014.psd, Google Drive file: bröllop 2.ai, Google Drive file: Bröllopsplaner.docx, Google Drive file: Schema, Google Drive file: Wedding (with Ninna) wine list] to "/home/yakir/.julia/datadeps/GoogleDriveDemo"?
```
but then it errors out:
```julia
INFO: [y/n]
y
WARNING: Could not resolve filename due to
WARNING: Could not resolve filename due to
WARNING: Could not resolve filename due to
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: Could not resolve filename due to
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: Could not resolve filename due to
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
WARNING: Could not resolve filename due to
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
ERROR: MethodError: Cannot `convert` an object of type GDFile to an object of type HTTP.URIs.URI
This may have arisen from a call to the constructor HTTP.URIs.URI(...),
since type constructors fall back to convert methods.
WARNING: falling back to using final part of remotepath
ERROR: WARNING: falling back to using final part of remotepath
LoadError: PyError (ccall(@pysym(:PyObject_Call), PyPtr, (PyPtr, PyPtr, PyPtr), o, arg, C_NULL)) <class 'pydrive.files.FileNotDownloadableError'>
FileNotDownloadableError('No downloadLink/exportLinks for mimetype found in metadata',)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 210, in GetContentFile
    self.FetchContent(mimetype, remove_bom)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 43, in _decorated
    return decoratee(self, *args, **kwargs)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 265, in FetchContent
    'No downloadLink/exportLinks for mimetype found in metadata')

Stacktrace:
 [1] #573 at ./asyncmap.jl:178 [inlined]
 [2] foreach(::Base.##573#575, ::Array{Any,1}) at ./abstractarray.jl:1733
 [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:178
 [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Int64, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:154
 [5] #async_usemap#558(::Int64, ::Void, ::Function, ::DataDeps.##20#21{#drive_download,String}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:103
 [6] (::Base.#kw##async_usemap)(::Array{Any,1}, ::Base.#async_usemap, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./<missing>:0
 [7] #download#19(::Array{GDFile,1}, ::Void, ::Bool, ::Function, ::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:48
 [8] handle_missing(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:6
 [9] _resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:83
 [10] resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:31
 [11] resolve(::String, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:54
 [12] resolve(::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:73
 [13] include_from_node1(::String) at ./loading.jl:576
 [14] include(::String) at ./sysimg.jl:14
while loading /home/yakir/pydrive.jl, in expression starting on line 13
```

and btw, I'm on DataDeps 0.3.5 (not master).

**Comment by oxinabox** [comment]
hmmm, yes, idk what is with that.
I could debug
But it doesn't matter really
because the version I just pushed into this repo works fine with DataDeps v0.4.0
(which is the name for what is master right now, but will not be for long since i am switching to 0.7)

**Comment by yakir12** [comment]
Awesome news about the 0.7. But I can't seem to find DataDeps v0.4.0 to try this with:
```julia
julia> Pkg.available("DataDeps")
10-element Array{VersionNumber,1}:
 v"0.1.0"
 v"0.2.1"
 v"0.2.2"
 v"0.2.3"
 v"0.2.4"
 v"0.2.6"
 v"0.3.0"
 v"0.3.3"
 v"0.3.4"
 v"0.3.5"
```

**Comment by oxinabox** [comment]
Arg, sorry for the lack of reply. I must have missed the notification.

V0.4.0  is being slow to be registered.
You can use git to checkout the tag, til tgen
I'll poke METADATA pr

**Comment by yakir12** [comment]
@oxinabox, is DataDeps v0.4.0 meant for Julia 0.7 only or is it supposed to work on Julia 0.6 as well?

**Comment by yakir12** [comment]
Humbly pinging you @oxinabox just because I saw you respond to one of my own Discourse posts and because I suspect the answer to my question is fairly straightforward? 

Apologies if this was unwarranted or bothersome.

**Comment by oxinabox** [comment]
no please, ping me more. 
 Something is up with my github and I don't seem to be getting notifications from this thread.

v0.4.0 should be the final 0.6.x release.
(well it might have later hotfixes as v0.4.1 etc)


Master has is for julia 0.7 only.
And soon v0.5.0 will be tagged off that

**Comment by yakir12** [comment]
cool, thanks.

**Comment by yakir12** [comment]
Finally managed to get v0.4.0, but still: 
```julia
julia> include("pydrive.jl")
INFO: This program has requested access to the data dependency GoogleDriveDemo.
INFO: which is not currently installed. It can be installed automatically, and you will not see this message again.
INFO: 
Demonstration of google drive

INFO: Do you want to download the dataset from GDFile[Google Drive file: Budget, Google Drive file: Bordsplacering, Google Drive file: Inköp, Google Drive file: Inköpslista:, Google Drive file: Ansvarsfördelning, Google Drive file: Lyrics Sånghäfte, Google Drive file: Skylt2.pdf, Google Drive file: Skylt.pdf, Google Drive file: Guest List, Google Drive file: Dekoration, Google Drive file: Ansvarsfördelning mat, Google Drive file: Ceremoni, Google Drive file: Adresslista, Google Drive file: Guest List with colors, Google Drive file: Invitation, Google Drive file: Copy of berg-bengtsson, Google Drive file: kommer ni eller inte?!, Google Drive file: IMG_7020.psd, Google Drive file: IMG_7014.psd, Google Drive file: bröllop 2.ai, Google Drive file: Bröllopsplaner.docx, Google Drive file: Schema, Google Drive file: Wedding (with Ninna) wine list] to "/home/yakir/.julia/datadeps/GoogleDriveDemo"?
INFO: [y/n]
y
ERROR: LoadError: PyError (ccall(@pysym(:PyObject_Call), PyPtr, (PyPtr, PyPtr, PyPtr), o, arg, C_NULL)) <class 'pydrive.files.FileNotDownloadableError'>
FileNotDownloadableError('No downloadLink/exportLinks for mimetype found in metadata',)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 210, in GetContentFile
    self.FetchContent(mimetype, remove_bom)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 43, in _decorated
    return decoratee(self, *args, **kwargs)
  File "/home/yakir/.julia/v0.6/Conda/deps/usr/lib/python2.7/site-packages/pydrive/files.py", line 265, in FetchContent
    'No downloadLink/exportLinks for mimetype found in metadata')

Stacktrace:
 [1] #573 at ./asyncmap.jl:178 [inlined]
 [2] foreach(::Base.##573#575, ::Array{Any,1}) at ./abstractarray.jl:1733
 [3] maptwice(::Function, ::Channel{Any}, ::Array{Any,1}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:178
 [4] wrap_n_exec_twice(::Channel{Any}, ::Array{Any,1}, ::Int64, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:154
 [5] #async_usemap#558(::Int64, ::Void, ::Function, ::DataDeps.##20#21{#drive_download,String}, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./asyncmap.jl:103
 [6] (::Base.#kw##async_usemap)(::Array{Any,1}, ::Base.#async_usemap, ::Function, ::Array{GDFile,1}, ::Vararg{Array{GDFile,1},N} where N) at ./<missing>:0
 [7] #download#19(::Array{GDFile,1}, ::Void, ::Bool, ::Function, ::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:50
 [8] handle_missing(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution_automatic.jl:8
 [9] _resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:83
 [10] resolve(::DataDeps.DataDep{Void,Array{GDFile,1},#drive_download,Base.#identity}, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:31
 [11] resolve(::String, ::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:54
 [12] resolve(::String, ::String) at /home/yakir/.julia/v0.6/DataDeps/src/resolution.jl:73
 [13] include_from_node1(::String) at ./loading.jl:576
 [14] include(::String) at ./sysimg.jl:14
while loading /home/yakir/pydrive.jl, in expression starting on line 13
```

**Comment by yakir12** [comment]
ping @oxinabox

**Comment by oxinabox** [comment]
Right, I am guessing the files you are wanting to download are in a  Google Apps format?
Like a GoogleSheets spreadsheet, or a Google Doc?
If so I think I have a solution and will commit it sometime soon (poke me if I forget)

**Comment by oxinabox** [comment]
Here we go.
I have pushed changes to master to handle that case.

**Comment by yakir12** [comment]
Wow, thank you so much!!

I've managed to get it to work with some simple example files :champagne:. But when I tried it with "real data" it seemed to work with some of the files but not all. So some of the files downloaded, but then it errored out mid-way. After some fiddling I think I discovered the reason: there was a folder in the folder I was trying to access. And indeed, my use-case is such that we have many nested folders within folders... Let me know if I can do anything to help..! (other than trying this out)

**Comment by oxinabox** [comment]
I don't have time to work on this in the foreseable future (thesis is due in <6 weeks).

Sadly GoogleDrive's API doesn't support **export folder and zip**,
which would just make everything really easy.

I feel like the example code already there demonstrates what would be needed to handle this case.

There are 2 paths.
The easy one, and the hard one.


### Easy Path

The easy way discards all structure of folders.
One simply filters out any GDFiles with the mimetype of `"application/vnd.google-apps.folder"`.
from the results of `list_files_in_folder`.

I  am not 100% sure if this will result in the files from those folders being downloaded  anyway.
but without folder structure.
I think it might, but if not, then I think the parents selector in the query to `list_files` needs to be changed to `ancestor` or some such.

### Hard Path.
DataDeps.jl does support nested `remote` lists,
and corresponding nested `post_fetch` methods.
I'll let you read the docs on those.

So a method would be created that would define nesting of files to download according to  their folder structure.
Basically like  `list_files_in_folder` 
But it would also define a nested series of functions for the `post_fetch` method,
that would to `mkpath` and `mv` to recreate the folder structure.


:-)
Good luck.

**Comment by yakir12** [comment]
Thank you so much for all the time you've already put into this! I understand completely and wish you the best of luck in your dissertation (which I'm sure you'll pass with flying colors)!

I'll keep hacking at this and report back if I ever have any major breakthroughs, hopefully resulting in a package.
